File: src/modules/Pieces/Kinds/Queen/Queen.cpp
#pragma once

#include "Queen.h"

Queen::Queen(PieceParams* params):
    WithRookMoves(),
    WithBishopMoves(),
    WithAnalizeDirections(params)
{
    this->withAnalizeDirectionsImpl->setGameController(
        this->pieceImpl->getGameController()
    );
}

vector<RawPosition> Queen::getPossibleMoves() {
    RawPosition* rawPosition = this->pieceImpl->getRawPosition();

    StateController bishopState = this->withBishopMovesImpl->getBishopMoves(rawPosition);

    vector<RawPosition> bishopMoves = this->withAnalizeDirectionsImpl->analizeDirections(&bishopState);

    StateController rookState = this->withRookMovesImpl->getRookMoves(rawPosition, true);

    vector<RawPosition> rookMoves = this->withAnalizeDirectionsImpl->analizeDirections(&rookState);

    vector<RawPosition> queenMoves = vector<RawPosition>();

    for (RawPosition position: bishopMoves) {
        queenMoves.push_back(position);
    }

    for (RawPosition position: rookMoves) {
        queenMoves.push_back(position);
    }

    return queenMoves;
}

File: packages/interactive_2d/chess/chess_react/src/Sketches/ChessSketch/elements/Pieces/Kinds/Queen.ts
/* eslint-disable @typescript-eslint/no-explicit-any */
import { Piece } from "../Piece";
import { getBishopMoves } from "../Utils/getBishopMoves";
import { getRookMoves } from "../Utils/getRookMoves";
import { Position } from "../types";

export class Queen extends Piece {
    /**
     * @override
     */
    public getPossibleMoves(): Position[] {
        const rookMoves = getRookMoves.bind(this as any);
        const bishopMoves = getBishopMoves.bind(this);

        return [
            ...rookMoves(),
            ...bishopMoves(),
        ];
    }
}
